feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt

```diff
@@ -0,0 +1,207 @@
+        skus: Set<String>,
+        onReceive: ProductDetailsListCallback,
+        onError: PurchasesErrorCallback
+    ) {
+        userDataHandler.getUserData { userData ->
+            productDataHandler.getProductData(skus, userData.marketplace, onReceive, onError)
+        }
+    }
+
+    // endregion
+
+    override fun consumeAndSave(
+        shouldTryToConsume: Boolean,
+        purchase: PurchaseWrapper
+    ) {
+        if (purchase !is AmazonPurchaseWrapper) return
+
+        if (purchase.type == ProductType.UNKNOWN) return
+
+        // PENDING purchases should not be fulfilled
+        if (purchase.purchaseState != RevenueCatPurchaseState.PURCHASED) return
+
+        if (shouldTryToConsume) {
+            PurchasingService.notifyFulfillment(purchase.purchaseToken, FulfillmentResult.FULFILLED)
+        }
+
+        // TODO: add logic for Unavailable
+    }
+
+    override fun consumeAndSave(
+        shouldTryToConsume: Boolean,
+        purchase: PurchaseHistoryRecordWrapper
+    ) {
+        // TODO
+    }
+
+    override fun consumePurchase(
+        token: String,
+        onConsumed: (billingResult: BillingResult, purchaseToken: String) -> Unit
+    ) {
+        TODO("not implemented")
+    }
+
+    override fun acknowledge(
+        token: String,
+        onAcknowledged: (billingResult: BillingResult, purchaseToken: String) -> Unit
+    ) {
+        TODO("not implemented")
+    }
+
+    override fun findPurchaseInPurchaseHistory(
+        skuType: ProductType,
+        sku: String,
+        completion: (BillingResult, PurchaseHistoryRecordWrapper?) -> Unit
+    ) {
+        // TODO
+    }
+
+    override fun makePurchaseAsync(
+        activity: Activity,
+        appUserID: String,
+        productDetails: ProductDetails,
+        replaceSkuInfo: ReplaceSkuInfo?,
+        presentedOfferingIdentifier: String?
+    ) {
+        if (replaceSkuInfo != null) {
+            log(LogIntent.AMAZON_WARNING, AmazonStrings.PRODUCT_CHANGES_NOT_SUPPORTED)
+            return
+        }
+        purchaseHandler.purchase(
+            appUserID,
+            productDetails,
+            presentedOfferingIdentifier,
+            onSuccess = { receipt, userData ->
+                backend.postAmazonReceiptData(
+                    receipt.receiptId,
+                    appUserID,
+                    userData.userId,
+                    productDetails,
+                    onSuccessHandler = { response ->
+                        val termSku = response["termSku"] as String
+                        val amazonPurchaseWrapper = AmazonPurchaseWrapper(
+                            sku = termSku,
+                            containedReceipt = receipt,
+                            presentedOfferingIdentifier = presentedOfferingIdentifier,
+                            purchaseState = RevenueCatPurchaseState.PURCHASED
+                        )
+                        purchasesUpdatedListener?.onPurchasesUpdated(listOf(amazonPurchaseWrapper))
```

Comment:
maybe this could be extracted into a function so that we're not going three levels deep?

Reply:
blocks have a tendency to grow over time in my experience, and if we're doing a lot in nested blocks it gets unreadable rather quickly